src/printer.js

```diff
@@ -0,0 +1,69 @@
+const { concat, trim } = require("./prettier");
+
+const comments = require("./comments");
+const embed = require("./embed");
+const nodes = require("./nodes");
+
+// This is the generic node print function, used to convert any node in the AST
+// into its equivalent Doc representation.
+function printNode(path, opts, print) {
+  const { type, body } = path.getValue();
+
+  if (type in nodes) {
+    return nodes[type](path, opts, print);
+  }
+
+  if (type[0] === "@") {
+    return body;
+  }
+
+  const ast = JSON.stringify(body, null, 2);
+  throw new Error(`Unsupported node encountered: ${type}\n${ast}`);
+}
+
+const noComments = ["args", "args_add_block"];
+
+// Certain nodes are used more for organizational purposed than for actually
+// displaying content, so we tell prettier that we don't want comments attached
+// to them.
+function canAttachComment(node) {
+  return !noComments.includes(node.type);
+}
+
+const nestedContents = ["assoclist_from_args", "bare_assoc_hash", "undef"];
+
+// This function tells prettier how to recurse down our AST so that it can find
+// where it needs to attach the comments.
+function getCommentChildNodes(node) {
+  return nestedContents.includes(node.type) ? node.body[0] : node.body;
+}
+
+// This is the generic print function for any comment in the AST. It handles
+// both regular comments that begin with a # and embdoc comments, which are
+// surrounded by =begin..=end.
+function printComment(path, _opts) {
+  const comment = path.getValue();
+
+  if (comment.type === "comment") {
+    return `#${comment.value}`;
```

Comment:
Hi @kddeisz,

We are trying to help with the Comments issue. When we run the test `yarn jest test/js/comments.test.js -t "loop do blocks as the only"`, we see that `comment.printed` is `true` although it is not in the result. We didn't really see where it was set to true and also didn't get who's calling `printComment` - Can you give us a hint where that happens?

Reply:
The answer unfortunately is kind of complicated.

Do you want to maybe pair on this? I'm in EST and free for the rest of the day except 1-2. We could set up a google hangout? My email is in the commits.

Reply:
Cool thanks for the offer :) -- I can't continue to work on it today, but maybe I'll let you know in advance when we continue on it and would be perfect to schedule a hangout 👍